pkg/services/network/nsxt_vpc_provider.go

```diff
@@ -0,0 +1,187 @@
+type nsxtVPCNetworkProvider struct {
+	client client.Client
+}
+
+// NSXTVpcNetworkProvider returns an instance of nsx-vpc type network provider.
+func NSXTVpcNetworkProvider(client client.Client) services.NetworkProvider {
+	return &nsxtVPCNetworkProvider{
+		client: client,
+	}
+}
+
+func (vp *nsxtVPCNetworkProvider) HasLoadBalancer() bool {
+	return true
+}
+
+func (vp *nsxtVPCNetworkProvider) SupportsVMReadinessProbe() bool {
+	// Note: The control plane VM network is private with nsx-vpc and
+	// readiness probe would fail. Therefore, nsxvpcNetworkProvider
+	// doesn't support VM readiness probe.
+	return false
+}
+
+// verifyNsxtVpcSubnetSetStatus checks the status conditions of a given SubnetSet within a cluster context.
+// If the subnet isn't ready, it is marked as false, and the function returns an error.
+// If the subnet is ready, the function updates the VSphereCluster with a "true" status and returns nil.
+func (vp *nsxtVPCNetworkProvider) verifyNsxtVpcSubnetSetStatus(vspherecluster *vmwarev1.VSphereCluster, subnetset *nsxopv1.SubnetSet) error {
+	clusterName := vspherecluster.Name
+	namespace := vspherecluster.Namespace
+	for _, condition := range subnetset.Status.Conditions {
+		if condition.Type == nsxopv1.Ready && condition.Status != corev1.ConditionTrue {
+			conditions.MarkFalse(vspherecluster, vmwarev1.ClusterNetworkReadyCondition, vmwarev1.ClusterNetworkProvisionFailedReason, clusterv1.ConditionSeverityWarning, condition.Message)
+			return errors.Errorf("subnetset ready status is: '%s' in cluster %s. reason: %s, message: %s",
+				condition.Status, types.NamespacedName{Namespace: namespace, Name: clusterName}, condition.Reason, condition.Message)
+		}
+	}
+
+	conditions.MarkTrue(vspherecluster, vmwarev1.ClusterNetworkReadyCondition)
+	return nil
+}
+
+// VerifyNetworkStatus checks if the given runtime object is of type SubnetSet.
+// If it is, then it calls verifyNsxVpcSubnetSetStatus with the SubnetSet to verify its status.
+// If it's not, it returns an error.
+func (vp *nsxtVPCNetworkProvider) VerifyNetworkStatus(_ context.Context, clusterCtx *vmware.ClusterContext, obj runtime.Object) error {
+	subnetset, ok := obj.(*nsxopv1.SubnetSet)
+	if !ok {
+		return fmt.Errorf("expected NSX VPC SubnetSet but got %T", obj)
+	}
+
+	return vp.verifyNsxtVpcSubnetSetStatus(clusterCtx.VSphereCluster, subnetset)
+}
+
+// ProvisionClusterNetwork provisions a new network in the context of a given cluster.
+// It constructs a new SubnetSet and attempts to create or patch it on the cluster.
+// If it fails to do so, it marks the status of the VSphereCluster as false and returns an error.
+// If it succeeds, it calls verifyNsxVpcSubnetSetStatus to verify the status of the newly created/patched SubnetSet.
+func (vp *nsxtVPCNetworkProvider) ProvisionClusterNetwork(ctx context.Context, clusterCtx *vmware.ClusterContext) error {
+	log := ctrl.LoggerFrom(ctx)
+
+	cluster := clusterCtx.VSphereCluster
```

Comment:
Let's do this instead please:
```go
	cluster := clusterCtx.VSphereCluster
	networkNamespace := cluster.Namespace
	networkName := cluster.Name
	log = log.WithValues("SubnetSet", klog.KRef(networkNamespace, networkName))

	log.Info("Provisioning ")
	defer log.Info("Finished provisioning")

	subnetset := &nsxopv1.SubnetSet{
		ObjectMeta: metav1.ObjectMeta{
			Namespace: networkNamespace,
			Name:      networkName,
		},
		Spec: nsxopv1.SubnetSetSpec{
			AdvancedConfig: nsxopv1.AdvancedConfig{
				StaticIPAllocation: nsxopv1.StaticIPAllocation{
					Enable: true,
				},
			},
		},
	}
```
(uses a proper k/v pair for logging, maps both namespace and name to network{Name,Namespace} not just one of them)

Reply:
Thanks for nice suggestion. Will addressed nsx-vpc part here and fix nsxt part in a separate pull request.